docs/src/literate/computational_homogenization.jl

```diff
@@ -0,0 +1,548 @@
+# returns the von Mise stress in every quadrature point for visualization.
+
+function compute_stress(cellvalues::CellVectorValues, dh::DofHandler, u, εᴹ)
+    σvM_qpdata = zeros(getnquadpoints(cellvalues), getncells(dh.grid))
+    σ̄Ω = zero(SymmetricTensor{2,2})
+    Ω = 0.0 # Total volume
+    for cell in CellIterator(dh)
+        E = cellid(cell) in getcellset(dh.grid, "inclusions") ? Ei : Em
+        reinit!(cellvalues, cell)
+        for q_point in 1:getnquadpoints(cellvalues)
+            dΩ = getdetJdV(cellvalues, q_point)
+            εμ = function_symmetric_gradient(cellvalues, q_point, u[celldofs(cell)])
+            σ = E ⊡ (εᴹ + εμ)
+            σvM_qpdata[q_point, cellid(cell)] = sqrt(3/2 * dev(σ) ⊡ dev(σ))
+            Ω += dΩ # Update total volume
+            σ̄Ω += σ * dΩ # Update integrated stress
+        end
+    end
+    σ̄ = σ̄Ω / Ω
+    return σvM_qpdata, σ̄
+end;
+
+# We now compute the homogenized stress and von Mise stress for all cases
+
+σ̄ = (
+    dirichlet = SymmetricTensor{2,2}[],
+    periodic  = SymmetricTensor{2,2}[],
+)
+σ = (
+     dirichlet = Vector{Float64}[],
+     periodic  = Vector{Float64}[],
+)
+
+projector = L2Projector(ip, grid)
+
+for i in 1:3
+    σ_qp, σ̄_i = compute_stress(cellvalues, dh, u.dirichlet[i], εᴹ[i])
+    proj = project(projector, σ_qp, qr; project_to_nodes=false)
+    push!(σ.dirichlet, proj)
+    push!(σ̄.dirichlet, σ̄_i)
+end
+
+for i in 1:3
+    σ_qp, σ̄_i = compute_stress(cellvalues, dh, u.periodic[i], εᴹ[i])
+    proj = project(projector, σ_qp, qr; project_to_nodes=false)
+    push!(σ.periodic, proj)
+    push!(σ̄.periodic, σ̄_i)
+end
+
+# The remaining thing is to compute the homogenized stiffness. As mentioned in the
+# introduction we can find all the components from the average stress of the sensitivity
+# fields that we have solved for
+#
+# ```math
+# \mathsf{E}_{ijkl} = \bar{\sigma}_{ij}(\hat{\boldsymbol{u}}_{kl}).
+# ```
+#
+# So we have now already computed all the components, and just need to gather the data in
+# a fourth order tensor:
+
+E_dirichlet = SymmetricTensor{4,2}((i, j, k, l) -> begin
+    if k == l == 1
+        σ̄.dirichlet[1][i, j] # ∂σ∂ε_**11
+    elseif k == l == 2
+        σ̄.dirichlet[2][i, j] # ∂σ∂ε_**22
+    else
+        σ̄.dirichlet[3][i, j] # ∂σ∂ε_**12 and ∂σ∂ε_**21
+    end
+end)
+
+E_periodic = SymmetricTensor{4,2}((i, j, k, l) -> begin
+    if k == l == 1
+        σ̄.periodic[1][i, j]
+    elseif k == l == 2
+        σ̄.periodic[2][i, j]
+    else
+        σ̄.periodic[3][i, j]
+    end
+end);
+
+# We can check that the result are what we expect, namely that the stiffness with Dirichlet
+# boundary conditions is higher than when using periodic boundary conditions, and that
+# the Reuss assumption is an lower bound, and the Voigt assumption a upper bound. We first
+# compute the Voigt and Reuss bounds:
+
+vm = 0.648 # Volume fraction for the matrix material
```

Comment:
Would it be nice to calculate the volume fraction automatically instead, should only be a few extra lines?

Reply:
Yea I could do that.